### ISSUE TITLE
Rollbar no longer working with Azure Functions.

### ISSUE THREAD

**Issue by opolo** [rollbar/Rollbar.NET#523]
Hi there!

Just joined Rollbar today as a user, so I have not experienced it working yet with Azure Functions (thus cannot say if Azure Functions used to work before). I wanted to use Rollbar to monitor the health of all of our Azure functions in our infrastructure, as they are moving data around without us knowing whether they are failing or not.

The issue is that whenever I create a RollbarConfig object, I get the error:
```
System.TypeInitializationException: 'The type initializer for 'Rollbar.PayloadStore.StoreContext' threw an exception.'

Inner Exception:
DllNotFoundException: Unable to load DLL 'e_sqlite3' or one of its dependencies: The specified module could not be found. (0x8007007E)
```

I believe this issue is related to the below linked issue combined with the fact that SQLLite usage cannot be avoided, when initializing the RollBarConfig class, even if e.g. the EnableLocalPayloadStore flag is set to false (I assume SQLLite is used, because it tries to read some local settings, which I figured that flag would disable):
https://github.com/Azure/azure-functions-host/issues/5597

Info: 
**Rollbar Client - Rollbar.NET Hosting Environment (please complete the following information):**
 - Rollbar.NET Host type: Running the Azure Function locally on our developer machines
 - OS and its version: W10
 - .NET flavor and its version: .NET Core 3.1
 - Application framework and its version: Azure Functions
 - Rollbar.NET Version: 3.10.0.
 - Applicable Rollbar Plug-in (if any): Rollbar.PlugIns.Serilog (The error happens before I pass the config object to the sink though)
 - Applicable custom third party logging library/framework integration (if any): Serilog
 - Applicable Browser (if any): N/A

**Additional context**
Will likely will have to go with another service-provider for this for now, as the feature needs to be up and running before our sprint ends this week, and I can imagine your release pipeline takes longer. This means though, you do not have to rush this bug fix request because of us. ...Just wanted to give a heads-up as Rollbar seems like a great (=simple to setup and use) product :).
If it can be fixed with some setting magic though, please let me know, as I don't want to research a new product in 1-2 days after just researching whether this could fulfill or needs :)

EDIT:
Image that shows how to produce error. The methods are called from within the Startup.cs -> Configure() method of my Function.
![image](https://user-images.githubusercontent.com/13970882/81077793-bb5c4080-8eed-11ea-97a3-1874a382e763.png)

**Comment by opolo** [comment]
Got it working by downgrading Rollbar Serilog Sink from 3.10.0 to 3.5.0 :D I don't think I have ever spend that much time refactoring to get a 3rd party library in (all logging setup was in a common class library, which caused some downgrade issues so had to split it all up)... It better have been worth it! Time to use it :D

Anyway, it still is an issue with any newer version, so I think this is an issue many new .NET users will face with your product, and unless they have 5h of Googling patience and experimentation, they might give up. We dont want that to happen! :)

**Comment by akornich** [comment]
@opolo, starting from v3.6.0 Rollbar.NET SDK uses sqlite database for its offline logging capabilities whenever there is any networking issue preventing the payloads from being sent to our API server. Azure Functions are not local-storage-friendly. So you can either use v3.5.x when hosting the SDK within an Azure Function or you can try using the latest version of the SDK while trying workaround described [here](https://stackoverflow.com/questions/49701806/why-does-using-sqlite-in-azure-functions-give-me-a-dllnotfoundexception).

**Comment by opolo** [comment]
@akornich Thanks for the response and feedback. 

Perhaps some option could be to have the logger be more "fire and forget"ísh, i.e. it tries to send the error, and if it fails, so be it... As a setting that is for e.g. Azure Functions as the logic does not work there anyway. In general I like the SQLite idea so nothing is lost though. 

I think your solution and/or the issue should be added to the documentation, despite their not being a good solution right now. As a newcomer that chooses Rollbar because of its simplicity "add these two lines and you can start sending logs" (although nearly everyone would want it more advanced, be it due to dependency injection or sinks into loggers), it provided an obstacle I can imagine would have quite a few bounce away from the product again sadly. I stayed and tried solving them, because a C# developer-friend recommended it though! :)

**Comment by akornich** [comment]
@opolo , thanks for the detailed feedback! it is much appreciated!

while the mainstream use/adoption of the SDK seems to be the case for most of our customers we do realize that there are a few rough edges when adopting the SDK within a couple of .NET "sub"-technologies, where Azure Functions is one of the two (another one is the SDK usage within Xamarin that depends on specific SDK hosting OS and its Sqlite dependency on it).

The SDK is already built based on the "fire and forget" based approach by default and the local payload persistence is optional. However, the main source of the problems you encountered and ones I just mentioned above is that even when you have the persistence turned off the dependencies on Sqlite modules are still there. Some of the modules are unmanaged/native modules and it looks like Azure Functions have trouble loading those (which sounds like by its current design, based on what I learned so far).

So, our current plan is to make persistence not just optional at run time but as an optional dependency that should be only available as an extra extension module to the core SDK component (the Rollbar package). The idea is that you would have to explicitly start using the persistence extension module with all its Sqlite dependencies if you do wish to enable the local payload persistence feature in your SDK hosting solution.

**Comment by opolo** [comment]
@akornich I think that would be a great way to solve it! It's a shame that the default for such a nice feature will be for it to be turned off, unless people install the extra dependency, but with some documentation I think that impact is minimal and most will see it. 

Solving it by adding having it in its own optional package, which can be added during the initialization of rollbar/the app, seems also very inline with the usual way .NET likes to solve things (ASP.NET Authentication, extensions, etc. comes to mind).

Tag me here, if you need someone to test it at a point, I can give it a spin in our Azure functions in our TEST environment and see how it goes.

**Comment by hanssens** [comment]
**TL;DR -** I can confirm having the same issues with a C# Azure Function, only during runtime, and also that downgrading to [v3.5.0](https://www.nuget.org/packages/Rollbar/3.5.0) fixes it.

I'm glad to hear it's on the team's radar to split-up external dependencies, like Sqlite, more gracefully in separate packages. 👍 +1 for me on both the error reporting and the proposed solution.

**Comment by akornich** [comment]
@hanssens, thank you for the feedback! Much appreciated!
